Tables refactor

sinkingsugar opened the pull request:
- Introduced a new member `fixedStructTable` in `SHTypeInfo` to optimize table handling.
- Updated `TableDecl` to mark tables as fixed struct tables when all keys are fixed values.
- Adjusted Rust type definitions to initialize `fixedStructTable` to false for consistency.


<!--
Before submitting your PR, please review the following checklist:

- [ ] **DO NOT** submit a PR without having discussed the change first in a related issue. If none exist, create one first.
- [ ] **CONSIDER** adding a unit test if your PR resolves an issue.
- [ ] **DO** keep pull requests small so they can be easily reviewed.
- [ ] **DO** make sure all unit tests pass.
- [ ] **DO** make sure not to introduce any new compiler warnings.
- [ ] **AVOID** breaking the continuous integration build.
- [ ] **AVOID** making significant changes to the overall architecture.
-->

sinkingsugar committed:
Add support for fixed struct tables in SHTypeInfo and TableDecl

- Introduced a new member `fixedStructTable` in `SHTypeInfo` to optimize table handling.
- Updated `TableDecl` to mark tables as fixed struct tables when all keys are fixed values.
- Adjusted Rust type definitions to initialize `fixedStructTable` to false for consistency.

sinkingsugar committed:
Add SHTableIndices and enhance fixed struct table handling

- Introduced `SHTableIndices` to optimize key access in tables.
- Updated `SHTableTypeInfo` to include `indices` and modified `fixedStructTable` handling.
- Implemented `deriveTableIndices` in `TableDecl` for efficient index generation.
- Added a new test case for fixed struct tables to ensure functionality.

sinkingsugar committed:
Enhance TableDecl with default value filling for table types

- Added `fillDefaultValues` method in `TableDecl` to initialize default values for table entries based on their types.
- Introduced new constructors for `Var` to support initialization with `SHTrait` and `SHTypeInfo`.
- Improved error handling in `TableOf` method for key and type length mismatches.

sinkingsugar committed:
Include numeric header in core.hpp for enhanced functionality

sinkingsugar committed:
Merge remote-tracking branch 'origin/devel' into gio/tables-refactor

sinkingsugar committed:
Refactor assert shard registration and enhance error handling in Set

- Renamed `Assert.Compose` and `Assert.NoCompose` to `Assert.ComposeSuccess` and `Assert.ComposeFailure` for clarity.
- Added error handling in `Set` to prevent usage of fixed struct tables, improving robustness.
- Updated test cases to reflect changes in assertion behavior.

sinkingsugar committed:
Enhance error handling in math operations and improve help text formatting

- Updated `dispatchType` function to include a `failed` parameter for better error handling in various mathematical operations.
- Modified help text formatting in `PlayShard` and `InterpolateShard` for improved readability.
- Added error checks in `Math` operations to handle invalid type scenarios gracefully.
- Refactored `operateDirect` and `operateBroadcast` methods across multiple shards to accommodate the new error handling mechanism.

sinkingsugar committed:
Enhance TableDecl and Clear error handling for fixed struct tables

- Added checks in `TableDecl` to skip filling default values for dynamic keys.
- Improved handling of `Clear` to prevent clearing fixed struct tables, enhancing robustness.
- Updated tests to assert failure scenarios for clearing fixed struct tables and invalid operations on struct entries.

sinkingsugar committed:
Enhance error handling for Erase and Clear operations on tables

- Added checks in `Erase` to ensure variables are mutable and to prevent erasing from fixed struct tables and tables with non-dynamic keys.
- Improved `Clear` to include similar checks for fixed struct tables and non-dynamic keys.
- Updated tests to assert failure scenarios for invalid erase and clear operations on tables.

sinkingsugar committed:
Implement dynamic table clearing logic in TableDecl

- Introduced a new member `_canClear` to determine if a table can be cleared based on its structure and key types.
- Updated the `Clear` logic to utilize `_canClear`, ensuring that only dynamic tables are cleared.
- Modified help text for the `Clear` parameter to clarify its behavior with dynamic tables.
- Adjusted tests to reflect changes in table declaration and clearing behavior.

sinkingsugar committed:
Refine Clear logic for dynamic table handling

- Enhanced the `Clear` operation to include checks for known keys, ensuring that only sequences or variable keys can be cleared.
- Improved error messages for invalid clear attempts on tables with fixed struct types or non-dynamic keys.
- Updated tests to validate the new behavior and error handling for dynamic tables.

sinkingsugar committed:
Refactor Erase operations for improved clarity and consistency

- Updated Erase examples in the documentation to enhance readability and maintain consistency in formatting.
- Added ExpectTable calls to ensure dynamic table handling is clear in examples.
- Reformatted log outputs for better alignment with expected results.

sinkingsugar committed:
Update variable assignment in table operations for clarity

- Changed the operation from `Set` to `Update` in the test for table declaration, improving the accuracy of the example.
- Ensured that the test reflects the intended behavior of updating values in a dynamic table.

guusw commented on shards/modules/core/core.hpp:
```diff
@@ -2614,6 +2658,16 @@ struct TableDecl : public VariableBase {
     }
 
     if (!_isTable) {
```
You should still check if type is `@type({none: something})`

guusw replied:
or has none key

sinkingsugar committed:
Add TODO for derived indices in TableDecl type handling

- Added a TODO comment in the `TableDecl` struct to indicate the need for utilizing derived indices in future optimizations of type handling.
- This change highlights an area for potential improvement in the type derivation process.
